README.rst
deploy.py
=========

.. image:: doc/twemproxy-sentinel-cluster.png

this script will deploy a redis cluster in **10 minutes**, with:

- redis
- redis-sentinel
- twemproxy

you can deploy/auto-failover/monitor/migrate and perform rolling-upgrades.

with all best practice from our experience (we use **1TB** of cache on production environment)

try it
------

0. prepare::

    sudo apt-get install git python python-pip
    pip install redis
    pip install -e git://github.com/idning/pcl.git#egg=pcl
    pip install -e git://github.com/kislyuk/argcomplete.git#egg=argcomplete
    git clone https://github.com/idning/redis-mgr.git

1. compile ``redis`` and ``twemproxy`` and put them under ``_binaries/`` dir::

    $ ll _binaries/
    total 19M
    1735820 -rwxr-xr-x 1 ning ning 705K 2014-03-24 19:26 nutcracker
    1735818 -rwxr-xr-x 1 ning ning 5.1M 2014-03-24 19:26 redis-sentinel
    1735819 -rwxr-xr-x 1 ning ning 5.1M 2014-03-24 19:26 redis-server
    1735815 -rwxr-xr-x 1 ning ning 3.8M 2014-03-24 19:26 redis-cli
    1735809 -rwxr-xr-x 1 ning ning  28K 2014-03-24 19:26 redis-check-aof
    1735801 -rwxr-xr-x 1 ning ning 3.7M 2014-03-24 19:26 redis-benchmark

2. choose your config filename::

    export REDIS_DEPLOY_CONFIG=conf && . bin/active

   you can use a private config file ``confconf_private.py`` and do ``export REDIS_DEPLOY_CONFIG=conf_private``

3. edit conf/conf.py

4. make sure you can ssh to target machine without having to input a password (use ``ssh-copy-id``)

5. run::

    $ ./bin/deploy.py cluster0 -h

monitor pages
-------------

(need http://www.google.com/jsapi)

.. image:: doc/redis-mgr-chart.png

config
------

::

    cluster0 = {
        'cluster_name': 'cluster0',
        'user': 'ning',
        'sentinel':[
            ('127.0.0.5:29001', '/tmp/r/sentinel-29001'),
            ('127.0.0.5:29002', '/tmp/r/sentinel-29002'),
            ('127.0.0.5:29003', '/tmp/r/sentinel-29003'),
        ],
        'redis': [
            # master(host:port, install path)       ,  slave(host:port, install path)
            ('127.0.0.5:20000', '/tmp/r/redis-20000'), ('127.0.0.5:21000', '/tmp/r/redis-21000'),
            ('127.0.0.5:20001', '/tmp/r/redis-20001'), ('127.0.0.5:21001', '/tmp/r/redis-21001'),
            ('127.0.0.5:20002', '/tmp/r/redis-20002'), ('127.0.0.5:21002', '/tmp/r/redis-21002'),
            ('127.0.0.5:20003', '/tmp/r/redis-20003'), ('127.0.0.5:21003', '/tmp/r/redis-21003'),
        ],
        'nutcracker': [
            ('127.0.0.5:22000', '/tmp/r/nutcracker-22000'),
            ('127.0.0.5:22001', '/tmp/r/nutcracker-22001'),
            ('127.0.0.5:22002', '/tmp/r/nutcracker-22002'),
        ],
    }

this will gen ``sentinel``  config::

    sentinel monitor cluster0-20000 127.0.0.5 20000 2
    sentinel down-after-milliseconds  cluster0-20000 60000
    sentinel failover-timeout cluster0-20000 180000
    sentinel parallel-syncs cluster0-20000 1

    sentinel monitor cluster0-20001 127.0.0.5 20001 2
    sentinel down-after-milliseconds  cluster0-20001 60000
    sentinel failover-timeout cluster0-20001 180000
    sentinel parallel-syncs cluster0-20001 1

and ``twemproxy`` config::

    cluster0:
      listen: 127.0.0.5:22000
      hash: fnv1a_64
      distribution: modula
      preconnect: true
      auto_eject_hosts: false
      redis: true
      backlog: 512
      client_connections: 0
      server_connections: 1
      server_retry_timeout: 2000
      server_failure_limit: 2
      servers:
        - 127.0.0.5:20000:1 cluster0-20000
        - 127.0.0.5:20001:1 cluster0-20001

the name ``cluster0-20000`` is named by the orig master,
if slave uses a different port, the server ``host:port``  of ``cluster0-20000`` can be ``127.0.0.5:20000`` or ``127.0.0.5:21000``

usage
-----

choose your config filename::

    export REDIS_DEPLOY_CONFIG=conf && . bin/active

::

    ning@ning-laptop:~/idning-github/redis-mgr$ ./bin/deploy.py cluster0 -h
    usage: deploy.py [-h] [-v] [-o LOGFILE] clustername op [cmd [cmd ...]]

    positional arguments:
      clustername           cluster0
      op                    migrate src dst : migrate a redis instance to another machine
                            web_server [port]: None
                            deploy          : deploy the binarys and config file (redis/sentinel/nutcracker) in this cluster
                            start           : start all instance(redis/sentinel/nutcracker) in this cluster
                            stop            : stop all instance(redis/sentinel/nutcracker) in this cluster
                            printcmd        : print the start/stop cmd of instance
                            status          : get status of all instance(redis/sentinel/nutcracker) in this cluster
                            log             : show log of all instance(redis/sentinel/nutcracker) in this cluster
                            rediscmd cmd    : run redis command against all redis instance, like 'INFO, GET xxxx'
                            mastercmd cmd   : run redis command against all redis Master instance, like 'INFO, GET xxxx'
                            rdb             : do rdb in all redis instance,
                            aof_rewrite     : do aof_rewrite in all redis instance
                            randomkill      : random kill master every mintue (for test failover)
                            sshcmd cmd      : ssh to target machine and run cmd
                            reconfigproxy   : sync the masters list from sentinel to proxy
                            failover        : catch failover event and update the proxy configuration
                            nbench [cnt]    : run benchmark against nutcracker
                            mbench [cnt]    : run benchmark against redis master
                            stopbench       : you will need this for stop benchmark
                            live_master_mem : monitor used_memory_human:1.53M of master
                            live_master_qps : monitor instantaneous_ops_per_sec of master
                            live_nutcracker_request : monitor nutcracker requests/s
                            live_nutcracker_forward_error : monitor nutcracker forward_error/s
                            live_nutcracker_inqueue : monitor nutcracker forward_error/s
                            live_nutcracker_outqueue : monitor nutcracker forward_error/s
                            live_overview [cnt]: overview monitor info of the cluster (from statlog file)
                            history [cnt]   : history monitor info of the cluster
                            upgrade_nutcracker : None
                            log_rotate      : log_rotate for nutcracker.
                            scheduler       : start following threads:
      cmd                   the redis/ssh cmd like "INFO"



these commands will affect the online running cluster status:

- start                 (only if master/slave connection is not running)
- stop                  (will ask for confirmation)
- reconfigproxy         (only if proxy config is out of date)
- randomkill            (will start it later)
- migrate

start cluster::

    $ ./bin/deploy.py cluster0 deploy

    $ ./bin/deploy.py cluster0 start
    2013-12-26 14:47:47,385 [MainThread] [NOTICE] start redis
    2013-12-26 14:47:47,622 [MainThread] [INFO] [redis:127.0.0.5:20000] start ok in 0.23 seconds
    2013-12-26 14:47:47,848 [MainThread] [INFO] [redis:127.0.0.5:21000] start ok in 0.22 seconds
    2013-12-26 14:47:48,099 [MainThread] [INFO] [redis:127.0.0.5:20001] start ok in 0.24 seconds
    2013-12-26 14:47:48,369 [MainThread] [INFO] [redis:127.0.0.5:21001] start ok in 0.27 seconds
    2013-12-26 14:47:50,788 [MainThread] [NOTICE] start sentinel
    2013-12-26 14:47:51,186 [MainThread] [INFO] [sentinel:127.0.0.5:29001] start ok in 0.39 seconds
    2013-12-26 14:47:51,452 [MainThread] [INFO] [sentinel:127.0.0.5:29002] start ok in 0.26 seconds
    2013-12-26 14:47:51,820 [MainThread] [INFO] [sentinel:127.0.0.5:29003] start ok in 0.35 seconds
    2013-12-26 14:47:51,820 [MainThread] [NOTICE] start nutcracker
    2013-12-26 14:47:52,082 [MainThread] [INFO] [nutcracker:127.0.0.5:22000] start ok in 0.26 seconds
    2013-12-26 14:47:52,364 [MainThread] [INFO] [nutcracker:127.0.0.5:22001] start ok in 0.28 seconds
    2013-12-26 14:47:52,573 [MainThread] [INFO] [nutcracker:127.0.0.5:22002] start ok in 0.21 seconds
    2013-12-26 14:47:52,573 [MainThread] [NOTICE] setup master->slave
    2013-12-26 14:47:52,580 [MainThread] [INFO] setup [redis:127.0.0.5:20000]->[redis:127.0.0.5:21000]
    2013-12-26 14:47:52,580 [MainThread] [INFO] [redis:127.0.0.5:21000] /home/ning/idning-github/redis/src/redis-cli -h 127.0.0.5 -p 21000 SLAVEOF 127.0.0.5 20000
    OK
    ...

run cmd on each master::

    $ ./bin/deploy.py cluster0 mastercmd 'get "hello"'
    2013-12-24 13:51:39,748 [MainThread] [INFO] [RedisServer:127.0.0.5:20000]: get "hello"
    [RedisServer:127.0.0.5:20000] xxxxx
    2013-12-24 13:51:39,752 [MainThread] [INFO] [RedisServer:127.0.0.5:20001]: get "hello"
    [RedisServer:127.0.0.5:20001]
    2013-12-24 13:51:39,756 [MainThread] [INFO] [RedisServer:127.0.0.5:20002]: get "hello"
    [RedisServer:127.0.0.5:20002]
    2013-12-24 13:51:39,760 [MainThread] [INFO] [RedisServer:127.0.0.5:20003]: get "hello"
    [RedisServer:127.0.0.5:20003] world

dump rdb for every redis instance::

    $ ./bin/deploy.py cluster0 rdb

monitor qps/memory::

    $ ./bin/deploy.py cluster0 mq
    2013-12-24 14:21:05,841 [MainThread] [INFO] start running: ./bin/deploy.py -v cluster0 mq
    2013-12-24 14:21:05,842 [MainThread] [INFO] Namespace(cmd=None, logfile='log/deploy.log', op='mq', target='cluster0', verbose=1)
    20000 20001 20002 20003
        6     5     5     6
        6     6     5     6
        6     6     5     6
     4741     6     6     6
    33106     5     5     6
    46639     8     7     7
    42265     6     5     7

run benchmark::

    $ ./bin/deploy.py cluster_offline0 bench
    $ ./bin/deploy.py cluster_offline0 mbench

modify config::

    $ ./bin/deploy.py cluster_offline0 mastercmd ' CONFIG GET save' -v
    $ ./bin/deploy.py cluster_offline0 mastercmd 'CONFIG SET save "10000 1000000"' -v

enable auto-complete
--------------------

::

    export REDIS_DEPLOY_CONFIG=conf

    pip install argcomplete
    $ . ./bin/active

    ning@ning-laptop ~/idning-github/redis-mgr$ ./bin/deploy.py cluster0 r<TAB>
    randomkill     rdb            reconfigproxy  rediscmd


gen_conf
--------

on ``bin/gen_conf.py`` use this ::

    BASEDIR = '/tmp/r'
    HOSTS = [
            '127.0.1.1',
            '127.0.1.2',
            '127.0.1.3',
            '127.0.1.4',
            ]
    MASTER_PER_MACHINE = 2
    SLAVE_PORT_INCREASE = 10000

it will gen the deploy.py config like this:

.. image:: doc/twemproxy-sentinel-cluster.png

migrate redis instance
----------------------

if we have 32 masters in 16 machines

1. dilatancy: move 2*32 instances on 16 machines to 32/64 machines (larger memory)
2. maintenance: one of the machines is down, we have to move data to another machine.

steps:

- pre_check,
- force_src_be_slave,
- deploy_dst,
- add_dst_as_slave,
- cleanup,
- sentinel_reset,
- update_config,

usage::

    $ ./bin/deploy.py cluster0 migrate cluster0-22000:127.0.0.5:23000:/tmp/r/redis-23000 cluster0-22000:127.0.0.5:50015:/tmp/r/redis-50015
    ...
    2014-02-27 19:21:58,667 [MainThread] [INFO] deploy [redis:127.0.0.5:50015]
    2014-02-27 19:21:59,774 [MainThread] [INFO] [redis:127.0.0.5:50015] start ok in 0.19 seconds
    2014-02-27 19:21:59,775 [MainThread] [NOTICE] add_dst_as_slave
    2014-02-27 19:21:59,790 [MainThread] [INFO] [redis:127.0.0.5:50015] /home/ning/idning-github/redis/src/redis-cli -h 127.0.0.5 -p 50015 SLAVEOF 127.0.0.5 22000
    OK
    2014-02-27 19:21:59,801 [MainThread] [INFO] [redis:127.0.0.5:50015]: {'used_memory': '342432', 'master_link_status': 'down', 'slave_repl_offset': '-1'}
    2014-02-27 19:22:00,811 [MainThread] [INFO] [redis:127.0.0.5:50015]: {'used_memory': '342464', 'master_link_status': 'down', 'slave_repl_offset': '-1'}
    2014-02-27 19:22:01,820 [MainThread] [INFO] [redis:127.0.0.5:50015]: {'used_memory': '363456', 'master_link_status': 'up', 'slave_repl_offset': '5998625'}
    2014-02-27 19:22:01,821 [MainThread] [NOTICE] cleanup
    2014-02-27 19:22:02,156 [MainThread] [INFO] [redis:127.0.0.5:23000] stop ok in 0.11 seconds
    2014-02-27 19:22:02,156 [MainThread] [NOTICE] sentinel_reset
    2014-02-27 19:22:02,165 [MainThread] [NOTICE] update_config
    2014-02-27 19:22:02,166 [MainThread] [INFO] AppendConfig:cluster0['migration'] = []
    2014-02-27 19:22:02,166 [MainThread] [INFO] AppendConfig:cluster0['migration'].append('cluster0-22000:127.0.0.5:23000:/tmp/r/redis-23000=>cluster0-22000:127.0.0.5:50015:/tmp/r/redis-50015')

this command will modify the conf.py::

    cluster0['migration'] = []
    cluster0['migration'].append('cluster0-22000:127.0.0.5:23000:/tmp/r/redis-23000=>cluster0-22000:127.0.0.5:50015:/tmp/r/redis-50015')

the "migration" section will auto load on next run::

    $ ./bin/deploy.py cluster0 status
    2014-02-27 19:24:24,815 [MainThread] [NOTICE] start running: ./bin/deploy.py -v cluster0 status
    2014-02-27 19:24:24,820 [MainThread] [NOTICE] status redis
    2014-02-27 19:24:24,825 [MainThread] [INFO] [redis:127.0.0.5:22000] uptime 29815 seconds
    2014-02-27 19:24:24,831 [MainThread] [INFO] [redis:127.0.0.5:50015] uptime 145 seconds
    ...
    2014-02-27 19:24:24,893 [MainThread] [NOTICE] status master-slave
    cluster0-22000 [redis:127.0.0.5:22000] <- 127.0.0.5:50015
    cluster0-22001 [redis:127.0.0.5:22001] <- 127.0.0.5:23001
    cluster0-22002 [redis:127.0.0.5:22002] <- 127.0.0.5:23002
    cluster0-22003 [redis:127.0.0.5:22003] <- 127.0.0.5:23003

mon as supervisor of twemproxy
------------------------------

mon: https://github.com/visionmedia/mon

this is optional for redis-mgr:

1. compile mon and put it in ``_binaries/``.
2. add config::

    BINARYS['MON_BINS'] = '_binaries/mon';

3. ./bin/deploy.py cluster0 upgrade_nutcracker

Dependencies
============

- `pcl <https://github.com/idning/pcl>`_
- `redis-py <https://github.com/andymccurdy/redis-py>`_ (<=2.9.0)
- `argcomplete <https://github.com/kislyuk/argcomplete>`_ (optional)
- `mon <https://github.com/visionmedia/mon>`_ (optional)
- if you are using python 2.7.3, you will need `this patch <http://bugs.python.org/msg158754>`_ to disable noise from threading

Authors
=======

- @idning
- @cen-li

TODO
====

1. scheduler for many clusters, we will need it! (we can use a shell script)
2. monitor ``SLOW LOG``
3. #live monitor for nutcracker
4. #nc to get nutcracker status will fail in background::

      nohup ./bin/deploy.py cluster0 scheduler  &

   we use telnetlib instead
5. migrate of redis instance
6. migrate data over cluster.
7. #a live command for cluster overview info(qps, mem, hit-rate)
8. make start cmd reentrant(slaveof cmd)
9. add ``max-mem`` config. on migration, makesure the max-mem config the same.
10. #upgrade nutcracker instance, support --filter
11. #add check_proxy_cfg

Graph
=====

- redis
    - mlive_mem
    - mlive_qps
- twemproxy
    - nlive_request
    - nlive_forward_error
    - nlive_inqueue
    - nlive_outqueue

- for cluster and for each instance
- support more than one cluster.
- do not need database


Changelog
- 2014-06-19 17:56:21
	- add chart.py, we can view qps/mem/forward_error_INC/hit_rate on chart.

- 2014-06-03 17:31:03
	- fix keys cmd(use sys.stdout.write insteadof print) 
	- use mon for supervisor of twemproxy
	- add live_master_slowlog/slowlog cmd, and we can monitor slowlog now
	- monitor threshold for each cluster 
	- use '-v 5' for first proxy in each cluster

- 2014-05-12 07:06:07
	- use hashtag {}
	- add keys cmd: ./bin/deploy.py cluster0 keys '*' | wc -l
	- add checkkcfg cmd(also in monitor task every minute)
	- add check_kv cmd(random incr kv)every minute
	- migrate slave work when slave down. 
	- config added: REDIS_MGR_CHECK_PREFIX = 'redis-mgr-check-'

- 2014-03-20 14:47:21

	https://github.com/idning/redis-mgr/compare/90cd823...3436805

	- new redis config format
	- add cmds: migrate, clean, sentinel_reset, sentinel_failover
	- cmd support more than one args (migrate a b)

- 2014-02-24 12:01:41
	- support more the one args(migrate a b)
	- migrate of redis instance




lib/server_modules.py
#!/usr/bin/env python
#coding: utf-8
#file   : server_modules.py
#author : ning
#date   : 2014-02-24 13:00:28


import os
import sys

from utils import *

class Base:
    '''
    the sub class should implement: _alive, _pre_deploy, status, and init self.args
    '''
    def __init__(self, name, user, host_port, path):
        self.args = {
            'name'      : name,
            'user'      : user,
            'host'      : socket.gethostbyname(host_port.split(':')[0]),
            'port'      : int(host_port.split(':')[1]),
            'path'      : path,

            'localdir'  : '',     #files to deploy

            'startcmd'  : '',     #startcmd and runcmd will used to generate the control script
            'runcmd'    : '',     #process name you see in `ps -aux`, we use this to generate stop cmd
            'logfile'   : '',
        }

    def __str__(self):
        return TT('[$name:$host:$port]', self.args)

    def deploy(self):
        logging.info('deploy %s' % self)
        self.args['localdir'] = TT('tmp/$name-$host-$port', self.args)
        self._run(TT('mkdir -p $localdir/bin && mkdir -p $localdir/conf && mkdir -p $localdir/log && mkdir -p $localdir/data ', self.args))

        self._pre_deploy()
        self._gen_control_script()
        self._init_dir()

        cmd = TT('rsync -ravP $localdir/ $user@$host:$path 1>/dev/null 2>/dev/null', self.args)
        self._run(cmd, timeout=5)

    def _gen_control_script(self):
        content = file('conf/control.sh').read()
        content = TT(content, self.args)

        control_filename = TT('${localdir}/${name}_control', self.args)

        fout = open(control_filename, 'w+')
        fout.write(content)
        fout.close()
        os.chmod(control_filename, 0755)

    def start(self, timeout = 60*5):
        if self._alive():
            logging.warn('%s already running' %(self) )
            return

        logging.debug('starting %s' % self)
        t1 = time.time()
        sleeptime = .5
        self._run(self._remote_start_cmd(), timeout=5) #timeout for run cmd

        while not self._alive():
            lets_sleep(sleeptime)
            if sleeptime < 5:
                sleeptime *= 2
            else:
                sleeptime = 5
                logging.warn('%s still not alive' % self)
            if time.time() - t1 > timeout:
                logging.warn('%s still not alive, Give Up !!! ' % self)
                logging.error('%s still not alive, Give Up !!! ' % self)
                return

        t2 = time.time()
        logging.info('%s start ok in %.2f seconds' %(self, t2-t1) )

    def stop(self, timeout = 15):
        if not self._alive():
            logging.warn('%s already stop' %(self) )
            return

        self._run(self._remote_stop_cmd(), timeout=5)
        t1 = time.time()
        while self._alive():
            lets_sleep()
            if time.time() - t1 > timeout:
                logging.warn('%s still not stop, Give Up !!! ' % self)
                logging.error('%s still not stop, Give Up !!! ' % self)
                return
        t2 = time.time()
        logging.info('%s stop ok in %.2f seconds' %(self, t2-t1) )

    def printcmd(self):
        print common.to_blue(self), self._remote_start_cmd()

    def status(self):
        logging.warn("status: not implement")

    def log(self):
        cmd = TT('tail $logfile', self.args)
        logging.info('log of %s' % self)
        print self._run(self._remote_cmd(cmd))

    def _sshcmd(self, cmd, timeout=60*60*24*30):
        '''
        run a benchmark cmd on this remote machine
        '''
        remote_cmd = self._remote_cmd(cmd)
        logging.info(remote_cmd)
        print self._run(remote_cmd, timeout)

    def _alive(self):
        logging.warn("_alive: not implement")

    def _init_dir(self):
        raw_cmd = TT('mkdir -p $path', self.args)
        self._run(self._remote_cmd(raw_cmd, chdir=False), timeout=5)

    def _remote_start_cmd(self):
        cmd = TT("./${name}_control start", self.args)
        return self._remote_cmd(cmd)

    def _remote_stop_cmd(self):
        cmd = TT("./${name}_control stop", self.args)
        return self._remote_cmd(cmd)

    def _remote_cmd(self, raw_cmd, chdir=True):
        if raw_cmd.find('"') >= 0:
            raise Exception('bad cmd: ' + raw_cmd)
        args = copy.deepcopy(self.args)
        args['cmd'] = raw_cmd
        if chdir:
            return TT('ssh -n -f $user@$host "cd $path && $cmd"', args)
        else:
            return TT('ssh -n -f $user@$host "$cmd"', args)

    def _run(self, raw_cmd, timeout=60*60*24*30):
        #ret = common.system(raw_cmd, logging.debug)
        ret = system_with_timeout(raw_cmd, logging.debug, timeout)
        if ret == None:
            logging.warn("cmd timeout: " + raw_cmd)
        ret = str(ret)
        logging.debug('return : [%d] [%s] ' % (len(ret), common.shorten(ret)) )
        return ret


class RedisServer(Base):
    def __init__(self, user, host_port, path):
        Base.__init__(self, 'redis', user, host_port, path)

        self.args['startcmd'] = TT('bin/redis-server conf/redis.conf', self.args)
        self.args['runcmd']   = TT('redis-server \*:$port', self.args)

        self.args['conf']     = TT('$path/conf/redis.conf', self.args)
        self.args['pidfile']  = TT('$path/log/redis.pid', self.args)
        self.args['logfile']  = TT('$path/log/redis.log', self.args)
        self.args['dir']      = TT('$path/data', self.args)

        self.args['REDIS_CLI'] = conf.BINARYS['REDIS_CLI']

    def _info_dict(self):
        cmd = TT('$REDIS_CLI -h $host -p $port INFO', self.args)
        info = self._run(cmd, timeout=5)

        info = [line.split(':', 1) for line in info.split('\r\n') if not line.startswith('#')]
        info = [i for i in info if len(i)>1]
        ret = defaultdict(str, info) #this is a defaultdict, be Notice
        try:
            ret['_slowlog_per_sec'] = self._slowlog_per_sec()
        except:
            pass
        return ret

    def _slowlog_per_sec(self):
        CNT = 10
        conn = redis.Redis(self.args['host'], self.args['port'])
        ret = conn.execute_command('SLOWLOG', 'GET', CNT)
        if not ret:
            return 0
        timediff = time.time() - ret[-1][1]  # time diff of the first-last slow log.
        slow_per_sec = 1.0 * CNT / timediff
        return slow_per_sec

    def _ping(self):
        cmd = TT('$REDIS_CLI -h $host -p $port PING', self.args)
        return self._run(cmd, timeout=5)

    def _alive(self):
        return strstr(self._ping(), 'PONG')

    def _gen_conf(self):
        content = file('conf/redis.conf').read()
        return TT(content, self.args)

    def _pre_deploy(self):
        self.args['BINS'] = conf.BINARYS['REDIS_SERVER_BINS']
        self._run(TT('cp $BINS $localdir/bin/', self.args))

        fout = open(TT('$localdir/conf/redis.conf', self.args), 'w+')
        fout.write(self._gen_conf())
        fout.close()

    def status(self):
        uptime = self._info_dict()['uptime_in_seconds']
        if uptime:
            logging.info('%s uptime %s seconds' % (self, uptime))
        else:
            logging.error('%s is down' % self)

    def isslaveof(self, master_host, master_port):
        info = self._info_dict()
        if info['master_host'] == master_host and int(info['master_port']) == master_port:
            logging.debug('already slave of %s:%s' % (master_host, master_port))
            return True

    def slaveof(self, master_host, master_port):
        cmd = 'SLAVEOF %s %s' % (master_host, master_port)
        return self.rediscmd(cmd)

    def rediscmd(self, cmd):
        args = copy.deepcopy(self.args)
        args['cmd'] = cmd
        cmd = TT('$REDIS_CLI -h $host -p $port $cmd', args)
        logging.info('%s %s' % (self, cmd))
        print self._run(cmd)


class Sentinel(RedisServer):
    def __init__(self, user, host_port, path, masters):
        RedisServer.__init__(self, user, host_port, path)

        self.args['startcmd'] = TT('bin/redis-sentinel conf/sentinel.conf', self.args)
        self.args['runcmd']   = TT('redis-sentinel \*:$port', self.args)

        self.args['conf']     = TT('$path/conf/sentinel.conf', self.args)
        self.args['pidfile']  = TT('$path/log/sentinel.pid', self.args)
        self.args['logfile']  = TT('$path/log/sentinel.log', self.args)

        self.args['name']     = 'sentinel'
        self.masters = masters

    def _gen_conf_section(self):
        template = '''\
sentinel monitor $server_name $host $port 2
sentinel down-after-milliseconds  $server_name 60000
sentinel failover-timeout $server_name 180000
sentinel parallel-syncs $server_name 1
        '''
        cfg = '\n'.join([TT(template, master.args) for master in self.masters])
        return cfg

    def _gen_conf(self):
        content = file('conf/sentinel.conf').read()
        content = TT(content, self.args)
        return content + self._gen_conf_section()

    def _pre_deploy(self):
        self.args['BINS'] = conf.BINARYS['REDIS_SENTINEL_BINS']
        self._run(TT('cp $BINS $localdir/bin/', self.args))

        fout = open(TT('$localdir/conf/sentinel.conf', self.args), 'w+')
        fout.write(self._gen_conf())
        fout.close()

    def get_masters(self):
        '''return currnet master list of (host:port, name)'''
        conn = redis.Redis(self.args['host'], self.args['port'])
        masters = conn.sentinel_masters()
        logging.debug('sentinel got masters: %s' % masters)
        return [('%s:%s' % (m['ip'], m['port']), m['name']) for m in masters.values()]

    def get_raw_masters(self):
        '''return currnet master list of (host:port, name)'''
        conn = redis.Redis(self.args['host'], self.args['port'])
        masters = conn.sentinel_masters()
        logging.debug('sentinel got masters: %s' % masters)
        return masters

    def get_raw_slaves(self, master_name):
        conn = redis.Redis(self.args['host'], self.args['port'])
        slaves = conn.sentinel_slaves(master_name)
        logging.debug('sentinel got slaves: %s' % slaves)
        return slaves

    def reset(self, master_name):
        '''
        reset all the masters with matching name.
        '''
        conn = redis.Redis(self.args['host'], self.args['port'])
        slaves = conn.sentinel('reset', master_name)
        logging.debug('sentinel reset')
        return slaves

    def failover(self, master_name):
        '''
        Force a failover as if the master was not reachable
        '''
        conn = redis.Redis(self.args['host'], self.args['port'])
        slaves = conn.sentinel('failover', master_name)
        logging.debug('sentinel failover %s' % master_name)
        return slaves

    def get_failover_event(self):
        self._sub = redis.Redis(self.args['host'], self.args['port']).pubsub()
        self._sub.subscribe('+switch-master')
        logging.info('subscribe +switch-master on %s' % self)
        iterator = self._sub.listen()
        if next(iterator)['channel'] != '+switch-master':
            raise Exception('error on subscribe')

        for msg in iterator:
            logging.info('got msg: %s' % msg)
            yield msg

class NutCracker(Base):
    def __init__(self, user, host_port, path, masters, verbose=4):
        Base.__init__(self, 'nutcracker', user, host_port, path)

        self.masters = masters

        self.args['conf']        = TT('$path/conf/nutcracker.conf', self.args)
        self.args['pidfile']     = TT('$path/log/nutcracker.pid', self.args)
        self.args['logfile']     = TT('$path/log/nutcracker.log', self.args)
        self.args['status_port'] = self.args['port'] + 1000
        self.args['verbose']     = verbose

        self.args['startcmd']    = TT('bin/nutcracker -d -c $conf -o $logfile -p $pidfile -s $status_port -v $verbose', self.args)
        self.args['runcmd']      = TT('bin/nutcracker -d -c $conf -o $logfile -p $pidfile -s $status_port', self.args)

        if 'MON_BINS' in conf.BINARYS and os.path.exists(conf.BINARYS['MON_BINS']):
            self.args['startcmd']    = TT('bin/mon -a 60 -d "bin/nutcracker -c $conf -o $logfile -p $pidfile -s $status_port -v $verbose"', self.args) # put -d at mon
            self.args['runcmd']      = TT('bin/nutcracker -c $conf -o $logfile -p $pidfile -s $status_port', self.args)

        self._last_info = None

    def _alive(self):
        return self._info_dict()

    def _gen_conf_section(self):
        template = '    - $host:$port:1 $server_name'
        cfg = '\n'.join([TT(template, master.args) for master in self.masters])
        return cfg

    def _gen_conf(self):
        content = '''
$cluster_name:
  listen: 0.0.0.0:$port
  hash: fnv1a_64
  distribution: modula
  preconnect: true
  auto_eject_hosts: false
  redis: true
  hash_tag: "{}"
  backlog: 512
  timeout: 400
  client_connections: 0
  server_connections: 1
  server_retry_timeout: 2000
  server_failure_limit: 2
  servers:
'''
        content = TT(content, self.args)
        return content + self._gen_conf_section()

    def _pre_deploy(self):
        self.args['BINS'] = conf.BINARYS['NUTCRACKER_BINS']
        self._run(TT('cp $BINS $localdir/bin/', self.args))
        if 'MON_BINS' in conf.BINARYS and os.path.exists(conf.BINARYS['MON_BINS']):
            self.args['MON_BINS'] = conf.BINARYS['MON_BINS']
            self._run(TT('cp $MON_BINS $localdir/bin/', self.args))

        fout = open(TT('$localdir/conf/nutcracker.conf', self.args), 'w+')
        fout.write(self._gen_conf())
        fout.close()

    def _info_dict(self):
        '''
                                                        | We will add fields in the info dict
        "uptime": 370,                                  |
        "timestamp": 1389231960,                        | timestamp_INC
        ....                                            |
        "cluster0": {                                   |
            "client_connections": 100,                  |
            "client_eof": 500,                          |
            "forward_error": 0,                         | calc forward_error_INC
            "client_err": 0,                            | calc client_err_INC
            "fragments": 0,                             |
            "server_ejects": 0,                         |
                                                        | add global in_queue/out_queue/
                                                        | add global requests/responses/
                                                        | add global server_timedout/server_err
                                                        | calc requests_INC responses_INC
                                                        | calc server_timedout_INC server_err_INC
            "cluster0-20001": {       #a backend        |
                "server_timedout": 0,                   |
                "server_err": 0,                        |
                "responses": 125406,                    |
                "response_bytes": 828478,               |
                "in_queue_bytes": 0,                    |
                "server_connections": 1,                |
                "request_bytes": 5189724,               |
                "out_queue": 0,                         |
                "server_eof": 0,                        |
                "requests": 125406,                     |
                "in_queue": 0,                          |
                "out_queue_bytes": 0                    |
            },                                          |
        '''
        info = self._raw_info_dict()
        #logging.debug(info)
        if not info:
            return None

        def calc_inc(cluster_name, info, last_info):
            TO_CALC_INC = ('forward_error', 'client_err', 'requests', 'responses', 'server_timedout', 'server_err')
            for item in TO_CALC_INC:
                info[item + '_INC'] = info[item] - last_info[item]

        def aggregation(cluster_name, info):
            TO_AGGREGATION = ('in_queue', 'out_queue', 'requests', 'responses', 'server_timedout', 'server_err')
            for item in TO_AGGREGATION:
                info[item] = 0
            for k, v, in info.items():
                if type(v) == dict: # a backend
                    for item in TO_AGGREGATION:
                        info[item] += v[item]

        if self._last_info:
            info['timestamp_INC'] = info['timestamp'] - self._last_info['timestamp']

        for k, v in info.items():
            if type(v) == dict:
                cluster_name = k
                cluster_info = v
                aggregation(cluster_name, cluster_info)
                if self._last_info:
                    calc_inc(cluster_name, cluster_info, self._last_info[cluster_name])

        self._last_info = info
        logging.debug(info)
        return info

    def _raw_info_dict(self):
        try:
            ret = telnetlib.Telnet(self.args['host'], self.args['status_port'], timeout=2).read_all()
            return common.json_decode(ret)
        except Exception, e:
            logging.debug('--- can not get _info_dict of nutcracker, [Exception: %s]' % (e, ))
            return None

    def status(self):
        ret = self._info_dict()
        if ret:
            uptime = ret['uptime']
            logging.info('%s uptime %s seconds' % (self, uptime))
        else:
            logging.error('%s is down' % self)

    def get_config(self):
        '''return currnet config file content, Ignore the listen: line'''
        cmd = TT('cat $conf', self.args)
        content = self._run(self._remote_cmd(cmd), timeout=5)
        content = re.sub('listen: .*', '', content)
        content = re.sub('Permanently added.*', '', content)
        return content.strip()

    def get_masters(self):
        '''return currnet master list of (host:port, name)'''
        cmd = TT('cat $conf', self.args)
        content = self._run(self._remote_cmd(cmd), timeout=5)
        logging.debug('current proxy config: %s' % content)

        def parse_line(line):
            _x, host_port_w, name = line.split()
            host, port, _w = host_port_w.split(':')
            return ('%s:%s' % (host, port), name)
        return [parse_line(line) for line in content.split('\n') if line.startswith('    -')]

    def reconfig(self, masters):
        self.masters = masters
        self.stop(timeout = 5)
        self.deploy()
        self.start(timeout = 5)
        logging.info('proxy %s:%s is updated' % (self.args['host'], self.args['port']))


# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4


